### PR TITLE
Add Stream Wrapper purge cache on delete

### PR DESCRIPTION
**Pull request by hanifn**
## Description

Add a filter to purge file from cache on delete.

Branched off of changes in #1061 

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).

### CODE REVIEW

**Comment by mjangda** [comment]
Note that for testing this on a sandbox, we should revert the patch we have that comments out the file path check.

File: files/class-vip-filesystem.php [hunk 1]
```diff
@@ -87,6 +89,8 @@ public function run() {
 	private function add_filters() {
 
 		add_filter( 'upload_dir', [ $this, 'filter_upload_dir' ], 10, 1 );
+		add_filter( 'wp_check_filetype_and_ext', array( $this, 'filter_filetype_check' ), 10, 4 );
+		add_filter( 'wp_delete_file', array( &$this, 'filter_delete_file' ), 20, 1 );
```
[the remark this answers is not in the document]
Comment (hanifn) [reply files/class-vip-filesystem.php]: Hmm.. I must have accidentally copied that from somewhere. I'll remove that

File: files/class-vip-filesystem.php [hunk 2]
```diff
@@ -129,4 +133,179 @@ public function filter_upload_dir( $params ) {
...
+	 * @since   1.0.0
+	 * @access  private
+	 *
+	 * @param   string      $file_path
+	 *
+	 * @return  string
+	 */
+	private function strip_query_args( $file_path ) {
+		$url      = wp_parse_url( $file_path );
+		$file_uri = sprintf( '/%s%s', $url['host'], $url['path'] );
```
[the remark this answers is not in the document]
Comment (hanifn) [reply files/class-vip-filesystem.php]: That's `wp_parse_url` converts the string `vip://wp-content/uploads/2019/1` into:
```
scheme => 'vip'
host => 'wp-content'
path => '/uploads/2019/1'
query => ''
```
So I'll need to add that `/` back into the beginning. Actually now that I think about it, this function doesn't just strip the query arguments. It also strips the `vip://` protocol. So it should probably be more accurately named `get_file_uri_path`?

File: files/class-vip-filesystem.php [hunk 3]
```diff
@@ -129,4 +133,179 @@ public function filter_upload_dir( $params ) {
...
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 *
+	 * @param   string  $file_uri
+	 */
+	private function purge_file_cache( $file_uri ) {
+		$invalidation_url = get_site_url();
+		if ( is_multisite() && ! ( is_main_network() && is_main_site() ) ) {
+			$invalidation_url .= '/sites/' . get_current_blog_id();
```
[the remark this answers is not in the document]
Comment (hanifn) [reply files/class-vip-filesystem.php]: Agreed. I will optimise.

File: files/class-vip-filesystem.php [hunk 4]
```diff
@@ -87,6 +89,20 @@ public function run() {
...
+	/**
+	 * Remove the registered filters.
+	 *
+	 * @since    1.0.0
+	 * @access   private
+	 */
+	private function remove_filters() {
+
+		remove_filter( 'upload_dir', [ $this, 'filter_upload_dir' ], 10 );
+		remove_filter( 'wp_check_filetype_and_ext', array( $this, 'filter_filetype_check' ), 10 );
```
Comment (mjangda) [inline files/class-vip-filesystem.php]:
```suggestion
		remove_filter( 'wp_check_filetype_and_ext', array( $this, 'filter_filetype_check' ), 10 );
		remove_filter( 'wp_delete_file', [ $this, 'filter_delete_file' ], 20 );
```

File: files/class-vip-filesystem.php [hunk 5]
```diff
@@ -129,4 +145,182 @@ public function filter_upload_dir( $params ) {
...
+		$this->purge_file_cache( $file_uri );
+
+		// Return empty value so that `wp_delete_file()` won't try to `unlink` again
+		return '';
+	}
+
+	/**
+	 * Remove duplicate uploads base directory from file path
+	 *
+	 * Some of the intermediate file paths have the uploads `basedir` occur twice so we will need to
```
Comment (mjangda) [inline files/class-vip-filesystem.php]: Do we know why the basedir is showing up twice? That sounds like a bug somewhere.

Comment (mjangda) [reply files/class-vip-filesystem.php]: Ah, it's probably the path_join (https://core.trac.wordpress.org/browser/trunk/src/wp-includes/post.php#L5483), which doesn't handle URIs with protocols: https://developer.wordpress.org/reference/functions/path_join/

**Comment by hanifn** [comment]
r126687-deploy

**Pull request merged**